### PR TITLE
PTypes 50: Add PType.canonical constructors

### PR DESCRIPTION
**Pull request by tpoterba**
When we remove requiredness from virtual types, requiredness will
live on in both encoded types and physical types. This will require
us to use these types to generate the canonical physical type without
extra missing bits. We will need to use this pattern in, for instance,
TableKeyBy, which may need to encode/decode to shuffle, and will
produce a constructable physical type to decode from.

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/types/physical/PType.scala [hunk 1]
```diff
@@ -114,6 +110,51 @@ object PType {
...
+      case t: TSet => PSet(canonical(t.elementType), t.required)
+      case t: TDict => PDict(canonical(t.keyType), canonical(t.valueType), t.required)
+      case t: TTuple => PTuple(t.types.map(canonical), t.required)
+      case t: TStruct => PStruct(t.fields.map(f => PField(f.name, canonical(f.typ), f.index)), t.required)
+      case t: TNDArray => PNDArray(canonical(t.elementType), t.required)
+    }
+  }
+
+  // currently identity
+  def canonical(t: PType): PType = {
```
Comment (cseed) [inline hail/src/main/scala/is/hail/expr/types/physical/PType.scala]: How is this not just `canonical(t.virtualType)`?

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/types/physical/PType.scala]: requiredness. Canonical PTypes will still have a notion of requiredness, and I want this to work when we strip requiredness out of virtual types.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/types/physical/PType.scala]: but currently they are the same

Comment (cseed) [reply hail/src/main/scala/is/hail/expr/types/physical/PType.scala]: Ah, got it, thanks for clarifying!

**Pull request merged**